templates/commands/render.go

```diff
@@ -0,0 +1,167 @@
+	f.StringVar(&cli.StringVar{
+		Name:    "source",
+		Aliases: []string{"s"},
+		Example: "helloworld@v1",
+		Target:  &r.flagSource,
+		Usage: `Required. The location of the template to be instantiated. Many forms are accepted. ` +
+			`"helloworld@v1" means "github.com/abcxyz/helloworld repo at revision v1; this is for a template owned by abcxyz. ` +
+			`"myorg/helloworld@v1" means github.com/myorg/helloworld repo at revision v1; this is for a template not owned by abcxyz but still on GitHub. ` +
+			`"mygithost.com/mygitrepo/helloworld@v1" is for a template in a remote git repo but not owned by abcxyz and not on GitHub. ` +
+			`"mylocaltarball.tgz" is for a template not in git but present on the local filesystem. ` +
+			`"http://example.com/myremotetarball.tgz" os for a non-Git template in a remote tarball.`,
+	})
+	f.StringVar(&cli.StringVar{
+		Name:    "spec",
+		Example: "path/to/spec.yaml",
+		Target:  &r.flagSpec,
+		Default: "./spec.yaml",
+		Usage:   "The path of the .yaml file within the unpacked template directory that specifies how the template is rendered.",
+	})
+	f.StringVar(&cli.StringVar{
+		Name:    "dest",
+		Aliases: []string{"d"},
+		Example: "/my/git/dir",
+		Target:  &r.flagDest,
+		Default: ".",
+		Usage:   "Required. The target directory in which to write the output files.",
+	})
+	f.StringVar(&cli.StringVar{
+		Name:    "git-protocol",
+		Example: "https",
+		Default: "https",
+		Target:  &r.flagGitProtocol,
+		Usage:   "Either ssh or https, the protocol for connecting to GitHub. Only used if the template source is GitHub.",
+	})
+	f.StringMapVar(&cli.StringMapVar{
+		Name:    "input",
+		Example: "foo=bar",
+		Target:  &r.flagInputs,
+		Usage:   "The key=val pairs of template values; may be repeated.",
+	})
+	f.StringVar(&cli.StringVar{
+		Name:    "log-level",
+		Example: "info",
+		Default: "warning",
+		Target:  &r.flagLogLevel,
+		Usage:   "How verbose to log; any of debug|info|warning|error.",
+	})
+	f.BoolVar(&cli.BoolVar{
+		Name:    "force-overwrite",
+		Target:  &r.flagForceOverwrite,
+		Default: false,
+		Usage:   "If an output file already exists in the destination, overwrite it instead of failing.",
+	})
+	f.BoolVar(&cli.BoolVar{
+		Name:    "keep-temp-dirs",
+		Target:  &r.flagKeepTempDirs,
+		Default: false,
+		Usage:   "Preserve the temp directories instead of deleting them normally.",
+	})
+
+	return set
+}
+
+func (r *Render) parseFlags(args []string) error {
+	if err := r.Flags().Parse(args); err != nil {
+		return fmt.Errorf("failed to parse flags: %w", err)
+	}
+
+	if r.flagSource == "" {
+		return fmt.Errorf("--source is required")
+	}
+
+	return nil
+}
+
+func (r *Render) Run(ctx context.Context, args []string) error {
+	if err := r.parseFlags(args); err != nil {
+		return err
+	}
+
+	if r.fs == nil { // allow filesystem interaction to be faked for testing
+		r.fs = os.DirFS("/").(fs.StatFS) //nolint:forcetypeassert // safe per docs: https://pkg.go.dev/os#DirFS
+	}
```

Comment:
Non-blocking, but this kind of logic always worries me. I would prefer if we _didn't_ save the result back on the command and instead creating a local variable:

```go
fs := r.testFS
if fs == nil {
  fs = os.DirFS("/").(fs.StatFS)
}
```

Alternatively, we could set `Render.fs` when we create the command:

```go
&Render{
  fs: os.DirFS("/").(fs.StatFS)
}
```

and then override it in tests:

```
render.fs = testFS
```

Reply:
```go
&Render{
  fs: os.DirFS("/").(fs.StatFS)
}
```
This won't work with the current packaging. The root command is still in `cmd/abc` when it initializes the `Render` command

Reply:
Then the local variable it is!

Reply:
Switched to using a local fs variable rather than having a mutable field, which I agree with.